### PR TITLE
PARQUET-1405: Fix writing statistics into DataPageHeader

### PR DESCRIPTION
**Pull request by majetideepak**

### CODE REVIEW

File: cpp/src/parquet/column_writer.cc [hunk 1]
```diff
@@ -598,17 +589,18 @@ int64_t ColumnWriterImpl::Close() {
...
+
     // Write stats only if the column has at least one row written
     // From parquet-mr
     // Don't write stats larger than the max size rather than truncating. The
     // rationale is that some engines may use the minimum value in the page as
     // the true minimum for aggregations and there is no way to mark that a
     // value has been truncated and is a lower bound and not in the page.
-    if (rows_written_ > 0 && chunk_statistics.is_set() &&
-        chunk_statistics.max_stat_length() <=
-            properties_->max_statistics_size(descr_->path())) {
```
Comment (majetideepak) [inline cpp/src/parquet/column_writer.cc]: Checking for min, max stat lengths here previously was incorrect. We ended up not writing stats at all if the lengths exceeded the `max_statistics_size`. Now, the logic does the right thing by avoid only the `min`, `max` values but still setting `null` and `distinct` counts.

**Comment by wesm** [comment]
WIll review this shortly. I'm undertaking a significant refactor of the statistics and comparator code to address the "extern template" issue and also to vectorize the comparators so I'll have to address a rebase conflict a bit

**Comment by wesm** [comment]
@majetideepak would you mind if we reviewed and tried to get in https://github.com/apache/arrow/pull/4233 first, since this will be easier to rebase (I can take care of the rebase)?

**Comment by majetideepak** [comment]
@wesm definitely makes sense to get #4233 in first. One of us can rebase this later.

**Comment by majetideepak** [comment]
@wesm I rebased this after #4233 got merged.

**Comment by wesm** [comment]
Thank you

File: cpp/src/parquet/column_writer-test.cc [hunk 2]
```diff
@@ -246,6 +246,18 @@ class TestPrimitiveWriter : public PrimitiveTypedTest<TestType> {
     return metadata_accessor->is_stats_set();
   }
 
+  void metadata_get_stats_has_min_max(bool& has_min, bool& has_max) {
```
Comment (wesm) [inline cpp/src/parquet/column_writer-test.cc]: nit: this should use pointers instead of references for the "out" arguments. I will fix

**Review by wesm — APPROVED** [review]
One nit that I'll fix quickly but otherwise LGTM

**Comment by dgrahn** [comment]
Has this made it into a release?

**Comment by wesm** [comment]
Yes, it's in the Arrow 0.14.x series.

**Comment by dgrahn** [comment]
Is there a way to read files affected by this issue?

**Comment by wesm** [comment]
Can we take this to dev@parquet.apache.org?

**Pull request closed without merging**